database/pebble/db.go

```diff
@@ -0,0 +1,463 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package pebble
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync/atomic"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/units"
+	"github.com/cockroachdb/pebble"
+	"github.com/cockroachdb/pebble/bloom"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// pebbleByteOverHead is the number of bytes of constant overhead that
+	// should be added to a batch size per operation.
+	pebbleByteOverHead = 8
+)
+
+var (
+	_ database.Database = (*Database)(nil)
+	_ database.Batch    = (*batch)(nil)
+	_ database.Iterator = (*iter)(nil)
+
+	ErrInvalidOperation = errors.New("invalid operation")
+)
+
+type Database struct {
+	db     *pebble.DB
+	closed utils.Atomic[bool]
+	// closeCh is closed when Close() is called.
+	closeCh chan struct{}
+}
+
+type Config struct {
+	CacheSize                   int // Byte
+	BytesPerSync                int // Byte
+	WALBytesPerSync             int // Byte (0 disables)
+	MemTableStopWritesThreshold int // num tables
+	MemTableSize                int // Byte
+	MaxOpenFiles                int
+}
+
+func NewDefaultConfig() Config {
+	return Config{
+		CacheSize:                   units.GiB,
+		BytesPerSync:                units.MiB,
+		WALBytesPerSync:             units.MiB,
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                16 * units.MiB,
+		MaxOpenFiles:                4 * units.KiB,
+	}
+}
+
+func New(file string, cfg Config, log logging.Logger, _ string, _ prometheus.Registerer) (database.Database, error) {
+	// Original default settings are based on
+	// https://github.com/ethereum/go-ethereum/blob/release/1.11/ethdb/pebble/pebble.go
+	opts := &pebble.Options{
+		Cache:        pebble.NewCache(int64(cfg.CacheSize)),
+		BytesPerSync: cfg.BytesPerSync,
+		// Although we use `pebble.NoSync`, we still keep the WAL enabled. Pebble
+		// will fsync the WAL during shutdown and should ensure the db is
+		// recoverable if shutdown correctly.
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    runtime.NumCPU,
+		Levels:                      make([]pebble.LevelOptions, 7),
```

Comment:
Wondering if we should configure these levels like geth does instead of leaving these as defaults